README.md
# Node.js Forever Service Cookbook

This Chef cookbook sets up an Upstart or /etc/init.d service definition for one
or more Node.js scripts using the [Forever][0] service manager. Forever is
configured to restart each Node.js process should it ever fail.

## Cookbook Dependencies

This depends on the following cookbooks:

  * logrotate
  * n-and-nodejs

Note the use of [n-and-nodejs][1] rather than the more commonly used [nodejs][2]
cookbook. If using [Librarian-chef][3] to manage dependences, then include these
custom cookbooks in the project Cheffile as follows:

```
cookbook 'n-and-nodejs',
  :git => 'https://github.com/exratione/n-and-nodejs-cookbook',
  :ref => 'v0.0.2'

cookbook 'forever-service',
  :git => 'https://github.com/exratione/forever-service',
  :ref => 'v0.0.5'
```

## Usage

Include this cookbook in the Chef runlist after the installation of Node.js and
its dependencies:

```
run_list [
  'recipe[build-essential]',
  'recipe[curl]',
  'recipe[n-and-nodejs]',
  'recipe[forever-service]'
]
```

## Attributes

The example attributes below define two services running from the same script,
which is a common use case for a Node.js server application. Multiple running
processes improve redundancy and take advantage of the existence of multiple
processors.

Sane use of Forever requires that all concurrently running processes have unique
paths, so symlinks to the single Node.js script are used. If the processes have
to bind to one or more ports or take other distinct actions, then the Node.js
script can examine the unique symlink path used to invoke it and determine the
configuration it should use. For example, a script `server.js` might be
symlinked from `server-a.js`, `server-b.js`, and `server-c.js`, and from that
know that it must read from configuration file `a.json`, `b.json`, or `c.json`.

```
default_attributes(
  'forever-service' => {
    'forever' => {
      # Set which version of the Forever package is installed globally via NPM.
      'version' => '0.11.1'
    },
    'node' => {
      # This is added to the PATH environment variable in the service scripts.
      # It is the directory in which the node binary resides for a standard
      # installation.
      'bin' => '/usr/local/bin',
      # Used to set the NODE_PATH environment variable in the service scripts.
      'path' => '/usr/local/lib/node_modules',
    },
    'services' => {
      # The key 'nodejs-server-1' will be used as the service identifier.
      'nodejs-server-1' => {
        # A service description. If omitted this defaults to the service
        # identifier.
        'description' => 'Node.js server process 1.',
        # A service display name. If omitted this defaults to the service
        # identifier.
        'display_name' => 'Node.js Server 1',
        # If enable is set to false, this definition will not be used. This can
        # be useful in some more advanced uses of Chef wherein attributes are
        # layered and merged from different sources.
        'enable' => true,
        # Configuration used when invoking Forever in service scripts.
        'forever' => {
          # These two help prevent servers from thrashing if they fail
          # immediately on launch. See the Forever documentation for more
          # details.
          'min_uptime' => 5000,
          'spin_sleep_time' => 2000
        },
        # Logging configuration.
        'log' => {
          # Absolute path to the log file.
          'path' => '/var/log/node/server-1.log',
          # Set this to false if you don't want log rotation set up.
          'rotate' => {
            # The standard values for these logrotate configuration parameters
            # are accepted. They pass through to the logrotate cookbook.
            'frequency' => 'weekly',
            'rotate' => 52
          }
        },
        # Absolute path to the pidfile created by Forever.
        'pid_file_path' => '/var/run/node-server-1.pid',
        # Absolute path to the application start script.
        'start_script' => '/path/to/scripts/server.js',
        # Absolute path for a symlink to the application start script. This will
        # be created if not present. This attribute can be omitted, but note
        # that this usage of Forever requires that all services have unique
        # start script paths. When running multiple services from the same
        # script it is necessary to use symlinks to provide those unique paths.
        'start_script_symlink' => '/path/to/scripts/server-1.js',
        # If true start the service in the provisioning process, otherwise leave
        # it unstarted. This defaults to true if omitted.
        'start' => true,
        # Allowed values are 'upstart' and 'initd', defining which type of
        # service is set up.
        'type' => 'initd',
        # The name of the user that the service runs under. The service will
        # always launch as root, but it is expected that it will downgrade its
        # permissions run as this user.
        'user' => 'node'
      },
      # This is a clone of the service definition above, replacing 1 with 2
      # in the attributes where appropriate. See the comments above.
      'nodejs-server-2' => {
        'description' => 'Node.js server process 2.',
        'display_name' => 'Node.js Server 2',
        'enable' => true,
        'forever' => {
          'min_uptime' => 5000,
          'spin_sleep_time' => 2000
        },
        'log' => {
          'path' => '/var/log/node/server-2.log',
          'rotate' => {
            'frequency' => 'weekly',
            'rotate' => 52
          }
        },
        'pid_file_path' => '/var/run/node-server-2.pid',
        'start_script' => '/path/to/scripts/server.js',
        'start_script_symlink' => '/path/to/scripts/server-2.js',
        'start' => true,
        'type' => 'initd',
        'user' => 'node'
      }
    }
  },
  # The n-and-nodejs cookbook values.
  'n-and-nodejs' => {
    'n' => {
      'version' => '1.2.1'
    },
    'nodejs' => {
      'version' => 'stable'
    }
  }
)
```

## Upstart or Init.d

Choose which service type to set up for each service defined by setting the
`forever-service['SERVICE-IDENTIFIER']['type']` attribute to either `upstart` or
`initd`.

The init.d script template used here should work just fine in both RPM-based and
Debian-based distributions. The Upstart script requires Upstart to be present,
which is not the case for all Linux distributions.

## Service User and Permissions

The service launches the Node.js process as the root user. It is expected that
the Node.js script will downgrade permissions to the user specified in the
attributes after binding to privileged ports or performing whatever other
activities require root permissions. For example this is done as follows for
the `node` user:

```
process.setgid('node');
process.setuid('node');
```

This cookbook needs to know about the service user and create the user if it
doesn't already exist in order to correctly set up ownership and permissions
for the service log directory.

[0]: https://github.com/nodejitsu/forever
[1]: https://github.com/exratione/n-and-nodejs-cookbook
[2]: https://github.com/mdxp/nodejs-cookbook
[3]: https://github.com/applicationsonline/librarian-chef


CHANGELOG.md
# Changelog

## 0.0.5

  * Allow multiple services to be set up, and change attributes to support this.
  * Use n-and-nodejs rather than nodejs as a dependency.

## 0.0.4

  * Tidy up some odds and ends.
  * Fix a failure to start the init.d service in the provisioning process.
  * Add logrotate.

## 0.0.3

  * Add option not to start.

## 0.0.2

  * Fix Upstart service start.
  * Start rolled in to service recipes.

## 0.0.1

  * Initial release.


metadata.rb
name              'forever-service'
maintainer        'Reason'
maintainer_email  'reason@exratione.com'
license           'MIT'
description       'Set up one or more services to run designated Node.js scripts under the Forever process manager.'
version           '0.0.5'
recipe            'forever-service', 'Set up services to run designated Node.js scripts.'

depends 'n-and-nodejs'
depends 'logrotate'

%w{ fedora redhat centos amazon scientific oracle ubuntu debian }.each do |os|
  supports os
end

attribute 'forever-service/forever/version',
  :display_name => 'Forever version',
  :description => 'Version of the Forever process manager to install.',
  :default => '0.11.1'

attribute 'forever-service/node/bin',
  :display_name => 'Node binary path',
  :description => 'Absolute path to the directory containing Node.js binaries.',
  :default => '/usr/local/bin'

attribute 'forever-service/node/path',
  :display_name => 'NODE_PATH value',
  :description => 'Absolute path to the node modules directory.',
  :default => '/usr/local/lib/node_modules'

attribute 'forever-service/services',
  :display_name => 'Services.',
  :description => 'Defining parameters for each service to be created.',
  :required => 'required',
  :type => 'hash'


recipes/default.rb
#
# Install Node.js services running under Forever.
#

# Make sure that logrotate is available.
begin
  include_recipe 'logrotate'
rescue
  Chef::Log.error('The forever-service recipe requires the logrotate cookbook.')
end

# Install Forever globally.
execute "npm -g install forever@#{node['forever-service']['forever']['version']}"

# ----------------------------------------------------------------------------
# Loop through the service definitions.
# ----------------------------------------------------------------------------

node['forever-service']['services'].each do |identifier, definition|

  # --------------------------------------------------------------------------
  # Set some defaults.
  # --------------------------------------------------------------------------

  # Setting these defaults should help make errors a little more comprehensible.
  definition = {
    'description' => identifier,
    'display_name' => identifier,
    'enable' => false,
    'forever' => {
      'min_uptime' => 5000,
      'spin_sleep_time' => 2000
    },
    'log' => {
      'path' => false,
      'rotate' => {
        'frequency' => 'weekly',
        'rotate' => 52
      }
    },
    'pid_file_path' => '/var/run/#{identifier}.pid',
    'start_script' => '',
    'start_script_symlink' => false,
    'start' => true,
    'type' => 'initd',
    'user' => 'node'
  }.merge(definition)

  log "Examining service definition #{identifier}:"
  log definition

  # --------------------------------------------------------------------------
  # Skip this one if it is not set enabled.
  # --------------------------------------------------------------------------

  if !definition['enable']
    log "Service definition #{identifier} is not enabled. Skipping."
    next
  end

  # --------------------------------------------------------------------------
  # Make sure that the declared service user exists.
  # --------------------------------------------------------------------------

  # Note that the service launches as root: it is the responsibility of the
  # service to switch to using the declared user. We need to ensure that the
  # user exists so that we can correctly set up the log directory.
  user definition['user']

  # --------------------------------------------------------------------------
  # Set up the log directory.
  # --------------------------------------------------------------------------

  logdir = File.dirname(definition['log']['path'])
  directory logdir do
    action :create
    owner definition['user']
    group definition['user']
    mode 00775
    recursive true
    not_if { ::File.exists?(logdir) }
  end

  # --------------------------------------------------------------------------
  # Create a symlink to the start script.
  # --------------------------------------------------------------------------

  # By the way Forever works every service start script has to have a different
  # path in order to make things tractable in the service definitions.
  #
  # When running several identical processes from the same script it is thus
  # necessary to create symlinks to create different paths for Forever.
  if definition['start_script_symlink']
    link definition['start_script_symlink'] do
      to definition['start_script']
    end
  end

  # --------------------------------------------------------------------------
  # Install and start the service.
  # --------------------------------------------------------------------------

  # An init.d service.
  if definition['type'] == 'initd'
    service identifier do
      service_name identifier
      supports [:start, :stop, :restart, :status]
      action :nothing
    end

    template "/etc/init.d/#{identifier}" do
      source 'initd.erb'
      owner 'root'
      group 'root'
      mode 00755
      variables definition
      notifies :enable, "service[#{identifier}]"
      if definition['start']
        notifies :start, "service[#{identifier}]"
      end
    end

  # An upstart service.
  elsif definition['type'] == 'upstart'
    execute "start #{identifier}" do
      action :nothing
    end

    template "/etc/init/#{identifier}.conf" do
      source 'upstart.conf.erb'
      owner 'root'
      group 'root'
      mode 00644
      variables definition
      if definition['start']
        notifies :run, "execute[start #{identifier}]"
      end
    end

  # Invalid service type.
  else
    log "Allowed values for service type are: 'initd', 'upstart'." do
      level :error
    end
  end

  # --------------------------------------------------------------------------
  # Set up log rotation.
  # --------------------------------------------------------------------------

  if definition['log']['rotate']
    logrotate_app identifier do
      path definition['log']['path']
      frequency definition['log']['rotate']['frequency']
      rotate definition['log']['rotate']['rotate']
      options ['missingok', 'compress', 'notifempty']
    end
  end

  log "Completed service setup for #{identifier}."

end
